contracts/DependencyAudit.cdc

```diff
@@ -41,27 +41,99 @@ access(all) contract DependencyAudit {
+
+            emit UnstagedDependencies(dependencies: unstagedDependencies)
+        }
+    }
+
+    access(self) fun maybePanicOnUnstagedDependencies(_ unstagedDependencies: [Dependency]) {
+        // If `panicOnUnstaged` is set to false, the function will return without panicking
+        if !DependencyAudit.panicOnUnstaged {
+            return
+        }
+
+        // check if we should panic randomly
+        if !self.shouldPanicRandomly() {
+            return
+        }
+
+        var unstagedDependenciesString = ""
+        var numUnstagedDependencies = unstagedDependencies.length
+        var j = 0
+        while j < numUnstagedDependencies {
+            if j > 0 {
+                unstagedDependenciesString = unstagedDependenciesString.concat(", ")
             }
+            unstagedDependenciesString = unstagedDependenciesString.concat(unstagedDependencies[j].toString())
+
+            j = j + 1
+        }
+
+        // the transactions will fail with a message that looks like this: `error: panic: Found unstaged dependencies: A.2ceae959ed1a7e7a.MigrationContractStaging, A.2ceae959ed1a7e7a.DependencyAudit`
+        panic("This transaction is using dependencies not staged for Crescendo upgrade coming soon! Learn more: https://bit.ly/FLOWCRESCENDO. Dependencies not staged: ".concat(unstagedDependenciesString))
+    }
+
+    // shouldPanicRandomly is used to randomly panic on unstaged dependencies
+    // The probability of panicking is based on the current block height and the start and end block heights
+    // If the start block height is greater than or equal to the end block height, the function will return true
+    // The function will always return true if the current block is more than the end block height
+    // The function will always return false if the current block is less than the start block height
+    // The function will return true if a random number between the start and end block heights is less than the current block height
+    // This means the probability of panicking increases linearly as the current block height approaches the end block height
+    access(self) fun shouldPanicRandomly(): Bool {
+        // get start block height or true
+        // get end block height or true
+        // get current block height
+        // get random number between start and end
+        // if random number is less than current block return true
+        // else return false
+
+        let maybeBoundaries = self.getBoundaries()
+        if maybeBoundaries == nil {
+            // if settings are invalid use default behaviour: panic true
+            return true
+        }
+        let boundaries = maybeBoundaries!
+
+        let startBlock: UInt64 = boundaries.start
+        let endBlock: UInt64 = boundaries.end
+        let currentBlock: UInt64 = getCurrentBlock().height
+
+        if startBlock >= endBlock {
+            // if settings are invalid use default behaviour: panic true
```

Comment:
Couldn't we just avoid this case by adding input validation when setting boundaries?

Reply:
I added input validation, but also left this here just in case.